File: Cs/ScreenKeyboad/Form/Model/Keyboard.cs
using AYam.Common.ViewModel;
using System;
using System.Windows;

namespace AYam.ScreenKeyboad.Form.Model
{

    /// <summary>
    /// Keyboard.Model
    /// </summary>
    internal class Keyboard : IDisposable
    {

        #region Property

        /// <summary>
        /// 現在値
        /// </summary>
        public string NowValue;

        /// <summary>
        /// 入力値
        /// </summary>
        public string Text;

        /// <summary>
        /// 文字選択開始位置
        /// 今回未使用
        /// </summary>
        public int SelectionStart = 0;

        /// <summary>
        /// 選択文字数
        /// 今回未使用
        /// </summary>
        public int SelectionLength = 0;

        /// <summary>
        /// ボタンクリックコマンド
        /// </summary>
        public DelegateCommand<string> InputCommand;

        /// <summary>
        /// Shiftキー押下
        /// </summary>
        private bool _IsShift = false;

        /// <summary>
        /// Shiftキー押下プロパティ
        /// </summary>
        public bool IsShift
        {
            get { return _IsShift; }
            set
            {

                if (!IsCaps)
                {

                    _IsShift = value;

                    if (value)
                    {

                        ShiftOffVisible = Visibility.Hidden;
                        ShiftOnVisible = Visibility.Visible;

                    }
                    else
                    {

                        ShiftOffVisible = Visibility.Visible;
                        ShiftOnVisible = Visibility.Hidden;

                    }

                }

            }
        }

        /// <summary>
        /// Caps Lockキー押下
        /// </summary>
        public bool IsCaps = false;

        /// <summary>
        /// Shiftキー非押下時に表示
        /// </summary>
        public Visibility ShiftOffVisible = Visibility.Visible;

        /// <summary>
        /// Shiftキー押下時に表示
        /// </summary>
        public Visibility ShiftOnVisible = Visibility.Hidden;

        #endregion

        /// <summary>
        /// キーボード.ModelのBase
        /// </summary>
        /// <param name="value"></param>
        public Keyboard(string value)
        {
            Initialize(value);
        }

        /// <summary>
        /// 終了処理
        /// </summary>
        public void Dispose()
        { }

        /// <summary>
        /// 初期化
        /// </summary>
        /// <param name="nowValue">現在値</param>
        private void Initialize(string nowValue = "")
        {

            NowValue = nowValue;
            Text = "";
            SelectionStart = 0;
            SelectionLength = 0;
            IsShift = false;

        }

        /// <summary>
        /// 文字入力
        /// </summary>
        /// <param name="content">Button.Content</param>
        /// <returns>
        /// -1 : 通常処理
        ///  1 : 次フォーカスへ移動
        ///  2 : Shift処理
        /// </returns>
        public int InputText(string content)
        {

            int returnValue = -1;

            switch (content.ToUpper())
            {

                case "CLEAR":

                    Initialize();
                    break;

                case "BS":

                    // 先頭フォーカスは削除する文字がないため除外
                    if (SelectionStart > 0)
                    {
                        SelectionStart -= 1;
                        SelectionLength = 1;
                        ControlText();
                    }

                    IsShift = false;
                    break;

                case "±":

                    // ブランクでないか
                    if (Text.Length > 0)
                    {

                        // 先頭1文字が"-"か
                        if (Text.Substring(0, 1).Equals("-"))
                        {

                            Text = Text.Substring(1);

                            if (SelectionStart > 0)
                            {
                                SelectionStart -= 1;
                            }

                        }
                        else
                        {

                            Text = "-" + Text;
                            SelectionStart += 1;

                        }
                        SelectionLength = 0;

                    }
                    else
                    {

                        Text = "-";
                        SelectionStart = 1;
                        SelectionLength = 0;

                    }

                    IsShift = false;
                    break;

                case "NUMBERDOT":

                    if (Text.Length.Equals(0) 
                        || (Text.Length.Equals(1) 
                            && Text.Equals("-")
                            )
                        )
                    {
                        Text = Text + "0.";
                        SelectionStart = 2;
                        SelectionLength = 0;
                    }
                    else
                    {
                        ControlText(".");
                    }

                    IsShift = false;
                    break;

                case "00":

                    if (Text.Length.Equals(0) 
                        || (Text.Length.Equals(1) 
                            && (Text.Equals("-") || Text.Equals("0"))
                            )
                        || (Text.Length.Equals(2) && Text.Equals("-0"))
                        )
                    {
                        Text = Text.Contains("-") ? "-0" :  "0";
                        SelectionStart = Text.Length;
                        SelectionLength = 0;
                    }
                    else
                    {
                        ControlText("00");
                    }

                    IsShift = false;
                    break;

                case "NEXT":
                    returnValue = 1;
                    break;

                case "SHIFT":
                    IsShift = !IsShift;
                    returnValue = 2;
                    break;

                case "CAPS":

                    if (IsShift)
                    {
                        IsCaps = !IsCaps;
                    }
                    IsShift = false;

                    returnValue = 3;
                    break;

                default:
                    ControlText(content);
                    IsShift = false;
                    break;

            }

            return returnValue;

        }

        /// <summary>
        /// Textbox.Textから選択文字列を削除し、指定文字列を挿入して返す
        /// </summary>
        /// <param name="insertString">挿入文字列</param>
        /// /// <returns>Textbox.Textから選択文字列を削除し指定文字列を挿入した文字列</returns>
        protected virtual void ControlText(string insertString = "")
        {

            // 削除文字よりも前を取得
            string firstString = SelectionStart.Equals(0) ? "" : Text.Substring(0, SelectionStart);

            // 削除文字よりも後を取得
            string secondString = (SelectionStart + SelectionLength).Equals(Text.Length) ? "" : Text.Substring(SelectionStart + SelectionLength);

            // 文字結合
            Text = firstString + insertString + secondString;

            // 選択位置調整
            SelectionStart = firstString.Length + insertString.Length;
            SelectionLength = 0;

        }

    }

}

File: VB/ScreenKeyboard/ScreenKeyboard/Form/Model/Keyboard.vb
Imports AYam.Common.ViewModel
Imports System
Imports System.Windows

Namespace Form.Model

    ''' <summary>
    ''' Keyboard.Model
    ''' </summary>
    Public Class Keyboard
        Implements IDisposable

#Region "Property"

        ''' <summary>
        ''' 現在値
        ''' </summary>
        Public NowValue As String

        ''' <summary>
        ''' 入力値
        ''' </summary>
        Public Text As String

        ''' <summary>
        ''' 文字選択開始位置
        ''' 今回未使用
        ''' </summary>
        Public SelectionStart As Integer = 0

        ''' <summary>
        ''' 選択文字数
        ''' 今回未使用
        ''' </summary>
        Public SelectionLength As Integer = 0

        ''' <summary>
        ''' ボタンクリックコマンド
        ''' </summary>
        Public InputCommand As DelegateCommand(Of String)

        ''' <summary>
        ''' Shiftキー押下
        ''' </summary>
        Private _IsShift As Boolean = False

        ''' <summary>
        ''' Shiftキー押下プロパティ
        ''' </summary>
        Public Property IsShift As Boolean
            Get
                Return _IsShift
            End Get
            Set(value As Boolean)

                If Not IsCaps Then

                    _IsShift = value

                    If value Then

                        ShiftOffVisible = Visibility.Hidden
                        ShiftOnVisible = Visibility.Visible

                    Else

                        ShiftOffVisible = Visibility.Visible
                        ShiftOnVisible = Visibility.Hidden

                    End If

                End If

            End Set
        End Property

        ''' <summary>
        ''' Caps Lockキー押下
        ''' </summary>
        Public IsCaps As Boolean = False

        ''' <summary>
        ''' Shiftキー非押下時に表示
        ''' </summary>
        Public ShiftOffVisible As Visibility = Visibility.Visible

        ''' <summary>
        ''' Shiftキー押下時に表示
        ''' </summary>
        Public ShiftOnVisible As Visibility = Visibility.Visible


#End Region

        ''' <summary>
        ''' Keyboard.Model
        ''' </summary>
        ''' <param name="value">現在値</param>
        Public Sub New(value As String)

            NowValue = value
            Initialize(value)

        End Sub

        ''' <summary>
        ''' 終了処理
        ''' </summary>
        Public Sub Dispose() Implements IDisposable.Dispose

        End Sub

        ''' <summary>
        ''' 初期化
        ''' </summary>
        ''' <param name="nowValue">現在値</param>
        Private Sub Initialize(Optional nowValue As String = "")

            Text = nowValue
            SelectionStart = Text.Length
            SelectionLength = 0
            IsShift = False

        End Sub

        ''' <summary>
        ''' 文字入力
        ''' </summary>
        ''' <param name="content">Button.Content</param>
        ''' <returns>
        ''' -1 : 通常処理
        '''  1 : 次フォーカス移動
        '''  2 : Shift処理
        '''  3 : Caps処理
        ''' </returns>
        Public Function InputText(content As String) As Integer

            Dim returnValue As Integer = -1

            Select Case content.ToUpper()

                Case "CLEAR"

                    Initialize()

                Case "BS"

                    ' 先頭フォーカスは削除する文字がないため除外
                    If SelectionStart > 0 Then

                        SelectionStart -= 1
                        SelectionLength = 1
                        ControlText()

                    End If

                    IsShift = False

                Case "±"

                    ' ブランクでないか
                    If Text.Length > 0 Then

                        ' 先頭１文字が"-"か
                        If (Text.Substring(0, 1).Equals("-")) Then

                            Text = Text.Substring(1)

                            If (SelectionStart > 0) Then

                                SelectionStart -= 1

                            End If

                        Else

                            Text = "-" & Text
                            SelectionStart += 1

                        End If

                        SelectionLength = 0

                    Else

                        Text = "-"
                        SelectionStart = 1
                        SelectionLength = 0

                    End If

                    IsShift = False

                Case "NUMBERDOT"

                    If Text.Length.Equals(0) _
                        OrElse (Text.Length.Equals(1) AndAlso Text.Equals("-")) Then

                        Text = Text & "0."
                        SelectionStart = 2
                        SelectionLength = 0

                    Else

                        ControlText(".")

                    End If

                    IsShift = False

                Case "00"

                    If Text.Length.Equals(0) _
                        OrElse (Text.Length.Equals(1) AndAlso (Text.Equals("-") OrElse Text.Equals("0"))) _
                        OrElse (Text.Length.Equals(2) AndAlso Text.Equals("-0")) Then

                        Text = If(Text.Contains("-"), "-0", "0")
                        SelectionStart = Text.Length
                        SelectionLength = 0

                    Else

                        ControlText(content)

                    End If

                    IsShift = False

                Case "NEXT"

                    returnValue = 1

                Case "SHIFT"

                    IsShift = Not IsShift
                    returnValue = 2

                Case "CAPS"

                    If IsShift Then

                        IsCaps = Not IsCaps

                    End If

                    IsShift = False

                    returnValue = 3

                Case Else

                    ControlText(content)
                    IsShift = False

            End Select

            Return returnValue

        End Function

        ''' <summary>
        ''' Textbox.Textから選択文字列を削除し、指定文字列を挿入して返す
        ''' </summary>
        ''' <param name="insertString">挿入文字列</param>
        Private Sub ControlText(Optional insertString As String = "")

            ' 削除文字よりも前を取得
            Dim firstString As String = If(SelectionStart.Equals(0), "", Text.Substring(0, SelectionStart))

            ' 削除文字よりも後を取得
            Dim secondString As String = If((SelectionStart + SelectionLength).Equals(Text.Length), "", Text.Substring(SelectionStart + SelectionLength))

            ' 文字結合
            Text = firstString & insertString & secondString

            ' 選択位置調整
            SelectionStart = firstString.Length + insertString.Length
            SelectionLength = 0

        End Sub

    End Class

End Namespace
